src/Engine/ProtoCore/DSASM/Heap.cs

```diff
@@ -520,14 +570,262 @@ private void GCDisposeObject(StackValue svPtr, Executive exe)
             }
         }
 
+#if TRACING_GC
+        /// <summary>
+        /// Mark all items in the array.
+        /// </summary>
+        /// <param name="hp"></param>
+        /// <returns></returns>
+        private int TraverseArray(DSArray array)
+        {
+            var dict = array.ToDictionary();
+            int size = array.MemorySize;
+
+            foreach (var pair in array.ToDictionary())
+            {
+                var key = pair.Key;
+                if (key.IsReferenceType)
+                    size += PropagateMark(key);
+
+                var value = pair.Value;
+                if (value.IsReferenceType)
+                    size += PropagateMark(value);
+            }
+
+            return size;
+        }
+
+        private int TraverseObject(DSObject obj)
+        {
+            int size = obj.MemorySize;
+
+            foreach (var item in obj.VisibleItems)
+            {
+                if (item.IsReferenceType)
+                    size += PropagateMark(item);
+            }
+
+            return size;
+        }
+
+        /// <summary>
+        /// Recursively mark all objects referenced by the object and change the
+        /// color of this object to black.
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        private int PropagateMark(StackValue value)
```

Comment:
Is this PropagateMark or RecursiveMark?

If Propagate, what is being propagated?